Máster Universitario en Inteligencia Artificial  
Universidad Internacional de Valencia  
03MIAR: Algoritmos de Optimización  
📌 Daniel Felipe Hernández Mancipe

---

# 🛠 Proyecto: Trabajo Práctico

URL: [GitHub](https://github.com/danielhndz/03MIAR-Algoritmos-de-Optimizacion/blob/main/Project/Seminario_Algoritmos.ipynb)

### Problemas:

> 1. [ ] Sesiones de doblaje
> 2. [ ] Organizar los horarios de partidos de La Liga
> 3. [x] **Combinar cifras y operaciones**

### Descripción del problema:

- El problema consiste en **analizar** el siguiente problema y **diseñar** un algoritmo que lo **resuelva**.
- Disponemos de las 9 cifras ($c$) del 1 al 9 (excluimos el cero) y de los 4 signos ($s$) básicos de las operaciones fundamentales: suma ($+$), resta ($-$), multiplicación ($\cdot$), y división ($/$).
- Debemos **combinarlos alternativamente sin repetir ninguno de ellos** para obtener una cantidad dada. Un ejemplo sería para obtener el 4:
  $$4 + 2 - \frac{6}{3} \cdot 1 = 4$$

1. Debe analizarse el problema para encontrar todos los **valores enteros** posibles planteando las siguientes cuestiones:
   1. ¿Qué valor **máximo y mínimo** se pueden obtener según las condiciones del problema?

      **Matiz importante sobre el formato de la expresión**: hay dos interpretaciones válidas del enunciado.
      - **Forma estricta**: la expresión debe tener exactamente el patrón $d_1\,s_1\,d_2\,s_2\,d_3\,s_3\,d_4\,s_4\,d_5$ (empieza con cifra y alterna 5 cifras y 4 signos).
      - **Forma extendida**: además se permiten expresiones que empiezan con signo $-$ unario, por ejemplo $-d_1\,s_2\,d_2\,s_3\,d_3\,s_4\,d_4$.

      Con este matiz, el valor **máximo** coincide en ambos casos y el **mínimo** cambia en una unidad:

      **Caso A (forma estricta):**

      $$
      9 \cdot 8 + 7 - \frac{6}{3} = 72 + 7 - 2 = \boxed{77}
      $$

      $$
      1 + \frac{4}{2} - 8 \cdot 9 = 1 + 2 - 72 = \boxed{-69}
      $$

      **Caso B (forma extendida con signo $-$ unario inicial):**

      $$
      9 \cdot 8 + \frac{7}{1} - 2 = 72 + 7 - 2 = \boxed{77}
      $$

      $$
      -9 \cdot 8 / 1 + 2 = -72 + 2 = \boxed{-70}
      $$

   2. ¿Es posible encontrar **todos los valores enteros posibles** entre dicho mínimo y máximo?

      Sí, en ambos casos (estricto y extendido) se obtienen todos los enteros del intervalo entre el mínimo y el máximo.

      **Caso A (forma estricta):**
      - Mínimo = $-69$, máximo = $77$.
      - Tamaño del intervalo entero esperado:
        $$
        |[-69,77] \cap \mathbb{Z}| = 77 - (-69) + 1 = 147
        $$
      - Valores enteros únicos obtenidos por fuerza bruta: $147$.
      - Conclusión: **sí** se obtienen todos los enteros del intervalo.

      **Caso B (forma extendida con signo $-$ unario inicial):**
      - Mínimo = $-70$, máximo = $77$.
      - Tamaño del intervalo entero esperado:
        $$
        |[-70,77] \cap \mathbb{Z}| = 77 - (-70) + 1 = 148
        $$
      - Valores enteros únicos obtenidos por fuerza bruta: $148$.
      - Conclusión: **sí** se obtienen todos los enteros del intervalo.

(\*) La respuesta es obligatoria

2. (\*) ¿Cuantas posibilidades hay sin tener en cuenta las restricciones?

Sin las restricciones, se podrían combinar las cifras ($c$) y los signos ($s$) con repetición, es decir, en cada paso de la construcción de la expresión, se tienen las mismas posibilidades de elección tanto para cifras como para signos. Por lo tanto, el número de posibilidades sin tener en cuenta las restricciones es:

$$
c \cdot s \cdot c \cdot s \cdot c \cdot s \cdot c \cdot s \cdot c =
$$

$$
c^5 \cdot s^4 = 9^5 \cdot 4^4 = \boxed{15,116,544}
$$

3. ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones?

Con el mismo matiz de las preguntas 1.1 y 1.2, hay dos recuentos válidos según cómo se interprete la estructura de la expresión.

**Caso A (forma estricta: $d_1\,s_1\,d_2\,s_2\,d_3\,s_3\,d_4\,s_4\,d_5$):**

- Selección ordenada de 5 cifras distintas de 9: $P(9,5)=9\cdot8\cdot7\cdot6\cdot5$.
- Selección ordenada de los 4 signos sin repetición: $P(4,4)=4\cdot3\cdot2\cdot1$.

$$
N_A = P(9,5)\cdot P(4,4)
$$

$$
N_A = (9\cdot8\cdot7\cdot6\cdot5)\,(4\cdot3\cdot2\cdot1)=9!=\boxed{362,880}
$$

**Caso B (forma extendida, añadiendo expresiones con signo $-$ unario al inicio):**

Además del caso estricto, se agregan expresiones del tipo $-d_1\,s_2\,d_2\,s_3\,d_3\,s_4\,d_4$.

- Selección ordenada de 4 cifras distintas de 9: $P(9,4)=9\cdot8\cdot7\cdot6$.
- Permutación de los 3 signos restantes ($+,*,/$): $P(3,3)=3!=6$.

$$
N_{extra}=P(9,4)\cdot P(3,3)=9\cdot8\cdot7\cdot6\cdot3!=\boxed{18,144}
$$

$$
N_B = N_A + N_{extra}=362,880+18,144=\boxed{381,024}
$$

Por lo tanto:

- Si se usa la **forma estricta**, hay **362,880** posibilidades.
- Si se usa la **forma extendida**, hay **381,024** posibilidades.

Como referencia empírica del algoritmo de fuerza bruta:

- Caso A: $90,000$ expresiones con resultado entero.
- Caso B: $94,824$ expresiones con resultado entero.

### Modelo para el espacio de soluciones

4. (\*) ¿Cuál es la estructura de datos que mejor se adapta al problema? Arguméntalo. (Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguméntalo)

No hay una única estructura óptima para todo el problema; el código usa una **combinación** según la función de cada fase:

- **`set`** para los dominios base (`FIGURES`, `SIGNS`) y para control de duplicados puntuales (`starting_minus` en fuerza bruta).
  Se usa porque garantiza unicidad y permite consultas de pertenencia eficientes.

- **`tuple`** para estados inmutables y ordenados en backtracking (`rem_figures`, `rem_signs`) y para las permutaciones generadas por `itertools`.
  Es clave porque esos estados se usan como claves hashables en `lru_cache` (función `suffix_transforms`).

- **`dict`** como estructura central de resultados: `expresión -> valor entero`.
  Esta estructura permite conservar el vínculo entre cada expresión y su evaluación, y facilita consultas posteriores por valor objetivo.

- **`dict` adicional de metadatos** para búsquedas por objetivo en fuerza bruta (`BRUTE_FORCE_TARGET_CACHE`), con clave `(modo, target)`.
  Se usa para reutilizar el resultado de búsquedas específicas ya ejecutadas.

- **`list` preasignada de tokens** en backtracking (buffer mutable) para construir expresiones in-place durante la DFS.
  Reduce creación de objetos frente a concatenaciones sucesivas en cada nivel recursivo.

- **`str`** como representación final de cada expresión evaluable (entrada de `eval` en fuerza bruta y clave del diccionario de resultados).

En resumen, el modelo de espacio de soluciones en esta implementación es: generación combinatoria (permutaciones/DFS), estado parcial compacto (`tuple` + escalares), y almacenamiento final en `dict` de expresiones válidas con resultado entero.

### Según el modelo para el espacio de soluciones

5. (\*) ¿Cuál es la función objetivo?

En sentido estricto, este trabajo no está planteado como una optimización clásica de una sola función escalar, sino como un problema de **búsqueda exhaustiva / enumeración** de expresiones factibles.

Sea $x$ una expresión candidata. Su evaluación se define como:

$$
E(x)=d_1\;s_1\;d_2\;s_2\;d_3\;s_3\;d_4\;s_4\;d_5
$$

(y, en la variante extendida, también se admiten expresiones con $-$ unario inicial).

El conjunto factible $\mathcal{X}$ queda determinado por:

- Dígitos sin repetición: $d_i \in \{1,\dots,9\}$ y $d_i \neq d_j$ para $i\neq j$.
- Operadores sin repetición: $s_k \in \{+, -, \cdot, /\}$ y $s_k \neq s_\ell$ para $k\neq \ell$.
- Evaluación válida respetando precedencia estándar.
- Filtro de salida: conservar solo resultados enteros, $E(x) \in \mathbb{Z}$.

Por tanto, la “función objetivo” operativa del trabajo es **construir y evaluar** todas las $x\in\mathcal{X}$ para obtener el conjunto de resultados enteros:

$$
\mathcal{R}=\{E(x)\in\mathbb{Z}: x\in\mathcal{X}\}
$$

6. (\*) ¿Es un problema de maximización o minimización?

No en su formulación principal. Es un problema de **factibilidad + enumeración** (o CSP discreto), no de maximización/minimización directa.

- Se acepta o rechaza cada expresión según cumpla restricciones.
- El algoritmo recorre el espacio de búsqueda y agrega los resultados válidos.

Si se quisiera forzar una formulación de optimización, podría reinterpretarse como una **maximización de cobertura** (maximizar $|\mathcal{R}|$), pero ese no es el objetivo original del enunciado ni del algoritmo implementado.

### Fuerza bruta

7. Diseña un algoritmo para resolver el problema por fuerza bruta.

El algoritmo por fuerza bruta se implementó mediante funciones pequeñas con responsabilidades claras, para mejor legibilidad y mantenimiento:

- **Orquestación principal**: `brute_force_alg(...)` coordina la ejecución completa (modo, búsqueda, metadata, impresión y retorno).
- **Construcción de expresiones**:
  - `build_strict_expression(...)` genera expresiones en forma estricta.
  - `build_extended_unary_expression(...)` genera la rama extendida con `-` unario inicial.
- **Generación del espacio de búsqueda**: `iter_candidate_expressions(...)` recorre candidatas por permutación de signos, evitando duplicados de la rama extendida.
- **Núcleo de búsqueda**: `run_brute_force_search(...)` evalúa candidatas y detiene en el primer hit si hay `target`.
- **Evaluación atómica**: `add_expression(...)` evalúa una candidata, actualiza contador/resultados y marca `found_target` cuando aplica.
- **Resumen de ejecución**: `build_brute_force_metadata(...)` centraliza la construcción del bloque de métricas.
- **Capa de reporte (salida)**:
  - `is_target_run(...)`, `print_summary_block(...)`, `get_extreme_witnesses(...)`,
    `print_extreme_witnesses(...)`, `print_missing_values_report(...)`, `print_wrapper(...)` y
    `print_target_status(...)` separan la lógica de impresión en piezas pequeñas.
  - `print_results(...)` actúa como orquestador de esas funciones de reporte.

Con `strict_form=True` (**forma estricta**) se generan solo expresiones del tipo
$d_1\,s_1\,d_2\,s_2\,d_3\,s_3\,d_4\,s_4\,d_5$.
Con `strict_form=False` (**forma extendida**) se añade también la rama con `-` unario al inicio.

En ambos modos se conservan únicamente expresiones con resultado entero y se reportan conteos, extremos y cobertura del intervalo $[\min,\max]$.

Para comparar ambos escenarios en la misma ejecución se realizan dos llamadas:

- `BRUTE_FORCE_RESULTS_EXTENDED = brute_force_alg(strict_form=False)`
- `BRUTE_FORCE_RESULTS_STRICT = brute_force_alg(strict_form=True)`


In [1]:
import time
import itertools
import textwrap


# Number of random targets used in helper experiments.
LIMIT = 10
# Digits allowed by the statement (1..9, no repetition inside an expression).
FIGURES = set(range(1, 10))
# Arithmetic operators allowed by the statement
# (no repetition inside an expression).
SIGNS = {"+", "-", "*", "/"}
# Cache for target runs: key=(mode, target), value=metadata dict.
BRUTE_FORCE_TARGET_CACHE = {}


def get_mode_name(strict_form: bool) -> str:
    """Return mode label used in logs/metadata."""
    return "strict" if strict_form else "extended"


def build_strict_expression(
    figures_perm: tuple,
    signs_perm: tuple,
) -> str:
    """Build strict-form expression: d1 s1 d2 s2 d3 s3 d4 s4 d5."""
    tokens = []
    for i in range(4):
        tokens.append(str(figures_perm[i]))
        tokens.append(signs_perm[i])
    tokens.append(str(figures_perm[-1]))
    return "".join(tokens)


def build_extended_unary_expression(
    figures_perm: tuple,
    signs_perm: tuple,
) -> str:
    """Build extended expression with unary minus at the start.

    Assumes `signs_perm[0] == '-'` and uses only the first 4 digits from
    `figures_perm`, preserving the original generation semantics.
    """
    tokens = []
    for i in range(4):
        tokens.append(signs_perm[i])
        tokens.append(str(figures_perm[i]))
    return "".join(tokens)


def iter_candidate_expressions(
    signs_perm: tuple,
    strict_form: bool,
):
    """Yield candidate expressions for one operator permutation."""
    # In extended mode, avoid regenerating the same
    # 4-digit unary-minus branch.
    starting_minus = set()

    for figures_perm in itertools.permutations(FIGURES, 5):
        yield build_strict_expression(figures_perm, signs_perm)

        # Extended mode adds unary-minus expression when first sign is '-'.
        if strict_form or signs_perm[0] != "-":
            continue

        first_four_digits = figures_perm[:-1]
        if first_four_digits in starting_minus:
            continue

        starting_minus.add(first_four_digits)
        yield build_extended_unary_expression(figures_perm, signs_perm)


def run_brute_force_search(
    strict_form: bool = False,
    target=None,
):
    """Evaluate candidates until completion or first target hit."""
    counter = 0
    # Stores only expressions whose evaluated value is an integer.
    results = {}
    target_state = {
        "target_found": False,
        "target_expression": "",
        "evaluated_before_first_target": None,
    }

    for signs_perm in itertools.permutations(SIGNS, 4):
        for expression in iter_candidate_expressions(
            signs_perm,
            strict_form,
        ):
            counter, found_target, found_expression = add_expression(
                expression,
                results,
                counter,
                target,
            )

            if found_target:
                target_state["target_found"] = True
                target_state["target_expression"] = found_expression
                target_state["evaluated_before_first_target"] = counter
                return results, counter, target_state

    return results, counter, target_state


def build_brute_force_metadata(
    mode: str,
    target,
    counter: int,
    results: dict,
    target_state: dict,
) -> dict:
    """Build summary metadata for brute-force runs."""
    return {
        "mode": mode,
        "target": target,
        "target_found": target_state["target_found"],
        "target_expression": target_state["target_expression"],
        "evaluated_before_first_target": (
            target_state["evaluated_before_first_target"]),
        "evaluated_expressions": counter,
        "expressions_with_integer_result": len(results),
        "unique_integer_results": len(set(results.values()))
    }


def brute_force_alg(
    lim: int = LIMIT,
    strict_form: bool = False,
    target=None,
    verbose: bool = True,
    return_metadata: bool = False,
) -> dict | tuple[dict, dict]:
    """Run brute force search for strict/extended form.

    Parameters
    ----------
    lim : int
        Maximum number of missing values shown in non-target runs.
    strict_form : bool
        True -> evaluate only strict expressions
        (d1 s1 d2 s2 d3 s3 d4 s4 d5).
        False -> include the extended unary-minus starting form.
    target : int | None
        If provided, stop on first integer expression equal to target.
    verbose : bool
        Whether to print the run summary.
    return_metadata : bool
        If True, return (results, metadata). Otherwise return only results.

    Returns
    -------
    dict | tuple[dict, dict]
        Dictionary expression -> integer result, and optional metadata.
    """
    start = time.time()
    mode = get_mode_name(strict_form)

    results, counter, target_state = run_brute_force_search(
        strict_form=strict_form,
        target=target,
    )

    metadata = build_brute_force_metadata(
        mode=mode,
        target=target,
        counter=counter,
        results=results,
        target_state=target_state,
    )

    # Keep target-run stats so repeated target queries can reuse them.
    if target is not None:
        BRUTE_FORCE_TARGET_CACHE[(mode, target)] = metadata

    if verbose:
        print_results(start, counter, results, lim, mode, metadata)

    if return_metadata:
        return results, metadata
    return results


def add_expression(
    expression: str,
    results: dict,
    counter: int,
    target=None
) -> tuple[int, bool, str]:
    """Evaluate one candidate expression and update counters/results.

    Parameters
    ----------
    expression : str
        Candidate expression built from controlled digits/operators.
    results : dict
        Dictionary expression -> integer result for integer-valued
        expressions.
    counter : int
        Number of expressions evaluated so far.
    target : int | None
        If provided, mark target hit when the integer result matches it.

    Returns
    -------
    tuple[int, bool, str]
        Updated counter, whether target was found, and the evaluated
        expression (empty string when the expression is skipped as duplicate).
    """
    # Skip duplicates created by mixed generation paths.
    if expression in results:
        return counter, False, ""

    # Safe here because expression is built from controlled digits/operators.
    result = eval(expression)
    counter += 1

    found_target = False
    if result.is_integer():
        integer_result = int(result)
        results[expression] = integer_result
        if target is not None and integer_result == target:
            found_target = True

    return counter, found_target, expression


def is_target_run(metadata) -> bool:
    """Return whether current reporting corresponds to a target run."""
    return metadata is not None and metadata.get("target") is not None


def print_wrapper(text: str):
    """Print text with textwrap and remove trailing newline."""
    print(textwrap.dedent(text).removesuffix("\n"))


def print_summary_block(
    mode: str,
    run_time: float,
    counter: int,
    integer_results: int,
    unique_integer_results: int,
    target_run: bool,
):
    """Print the main summary block for brute-force runs."""
    evaluated_line = (
        "Evaluated expressions until first target hit (inclusive) = "
        f"{counter}"
        if target_run
        else f"Evaluated expressions = {counter}"
    )
    print_wrapper(f"""
        Brute-force algorithm results ({mode} form):
        Run time = {run_time:.3f} seconds
        {evaluated_line}
        Evaluated expressions with integer result = {integer_results}
        Evaluated expressions with unique integer result = {
            unique_integer_results
        }
    """)


def get_extreme_witnesses(results: dict):
    """Return min/max values and one witness expression for each extreme."""
    if not results:
        return None

    values_set = set(results.values())
    min_value = min(values_set)
    max_value = max(values_set)
    min_expression = next(k for k, v in results.items() if v == min_value)
    max_expression = next(k for k, v in results.items() if v == max_value)

    return values_set, min_value, min_expression, max_value, max_expression


def print_extreme_witnesses(
    min_expression: str,
    min_value: int,
    max_expression: str,
    max_value: int,
):
    """Print min/max witness expressions."""
    print(f"Minimum value = {min_expression} = {min_value}")
    print(f"Maximum value = {max_expression} = {max_value}")


def print_missing_values_report(
    values_set: set,
    min_value: int,
    max_value: int,
    lim: int,
):
    """Print range-coverage information for non-target runs."""
    missing_values = set(range(min_value, max_value + 1)) - values_set
    all_ints_between_min_max = "Yes" if len(missing_values) == 0 else "No"
    print((
        "Did it get all the integers between the min and max values? "
        f"{all_ints_between_min_max}"
    ))

    if len(missing_values) > 0:
        print_wrapper(f"""
            First {lim} missing values for the full range:
            {sorted(list(missing_values))[:lim]}
        """)


def print_target_status(metadata):
    """Print target-found status for target runs."""
    target = metadata["target"]
    if metadata["target_found"]:
        print((
            "First expression found for target "
            f"{target}: {metadata['target_expression']}"
        ))
    else:
        print(f"No expression found for target {target}.")


def print_results(
    start: float,
    counter: int,
    results: dict,
    lim: int = LIMIT,
    mode: str = "extended",
    metadata=None,
):
    """Print a human-readable summary for brute-force runs."""
    run_time = time.time() - start
    target_run = is_target_run(metadata)

    extreme_data = get_extreme_witnesses(results)
    values_set = extreme_data[0] if extreme_data is not None else set()

    print_summary_block(
        mode=mode,
        run_time=run_time,
        counter=counter,
        integer_results=len(results),
        unique_integer_results=len(values_set),
        target_run=target_run,
    )

    if extreme_data is not None:
        _, min_value, min_expression, max_value, max_expression = (
            extreme_data
        )
        print_extreme_witnesses(
            min_expression=min_expression,
            min_value=min_value,
            max_expression=max_expression,
            max_value=max_value,
        )

        if not target_run:
            print_missing_values_report(
                values_set=values_set,
                min_value=min_value,
                max_value=max_value,
                lim=lim,
            )

    if target_run:
        print_target_status(metadata)


# Precompute full search spaces once for both modes.
BRUTE_FORCE_RESULTS_EXTENDED = brute_force_alg(strict_form=False)
BRUTE_FORCE_RESULTS_STRICT = brute_force_alg(strict_form=True)


Brute-force algorithm results (extended form):
Run time = 1.592 seconds
Evaluated expressions = 381024
Evaluated expressions with integer result = 94824
Evaluated expressions with unique integer result = 148
Minimum value = -8/1*9+2 = -70
Maximum value = 7/1+8*9-2 = 77
Did it get all the integers between the min and max values? Yes

Brute-force algorithm results (strict form):
Run time = 1.490 seconds
Evaluated expressions = 362880
Evaluated expressions with integer result = 90000
Evaluated expressions with unique integer result = 147
Minimum value = 4/2+1-8*9 = -69
Maximum value = 7/1+8*9-2 = 77
Did it get all the integers between the min and max values? Yes


### Complejidad

7. Calcula la complejidad del algoritmo por fuerza bruta.

El coste del algoritmo por fuerza bruta se determina por dos factores:

- Cuántas expresiones se generan.
- Cuánto cuesta construir/evaluar cada expresión.

Definamos:

- $d$: número de cifras disponibles,
- $s$: número de signos disponibles,
- $m$: número de operadores usados en cada expresión (en este problema, $m=4$ y por tanto se usan $m+1=5$ cifras).

**Forma estricta** ($d_1\,s_1\,d_2\,s_2\,d_3\,s_3\,d_4\,s_4\,d_5$):

$$
N_{strict}=P(s,m)\cdot P(d,m+1)
$$

Para el caso del proyecto ($d=9$, $s=4$, $m=4$):

$$
N_{strict}=P(4,4)\cdot P(9,5)=4!\cdot \frac{9!}{(9-5)!}=\boxed{362{,}880}
$$

**Forma extendida** (añadiendo expresiones con $-$ unario inicial):

$$
N_{extra}=P(d,m)\cdot P(s-1,m-1)
$$

En este problema:

$$
N_{extra}=P(9,4)\cdot P(3,3)=\boxed{18{,}144}
$$

Luego:

$$
N_{extended}=N_{strict}+N_{extra}=\boxed{381{,}024}
$$

Si el coste de construir y evaluar una expresión es $O(m)$ (o $O(1)$ cuando $m$ es constante), la complejidad temporal queda:

$$
T_{strict}=O\big(P(s,m)\cdot P(d,m+1)\cdot m\big)
$$

$$
T_{extended}=O\big((P(s,m)\cdot P(d,m+1)+P(s-1,m-1)\cdot P(d,m))\cdot m\big)
$$

Para esta instancia concreta ($d=9$, $s=4$, $m=4$), $m$ es constante, así que el tiempo queda dominado por el número de expresiones evaluadas:

- Forma estricta: $\Theta(362{,}880)$,
- Forma extendida: $\Theta(381{,}024)$.

El coste en memoria es $O(k)$, donde $k$ es el número de expresiones con resultado entero que se almacenan en el diccionario de resultados.

### Búsqueda de una o todas las expresiones que evalúen a un objetivo

Ya con el espacio de soluciones completo, simplemente se tendría que buscar en el diccionario que claves tienen como valor el resultado de interés. Por ejemplo, se pueden obtener todas las expresiones que evalúan en `3` y `4` y la primera que evalúa a `42` y comparar ambos casos (estricto y extendido):


In [3]:
def parse_query_mode(mode: str) -> tuple[str, bool, dict]:
    """Parse mode input and return mode label, strict flag and results map."""
    normalized_mode = mode.lower()

    if normalized_mode == "strict":
        return normalized_mode, True, BRUTE_FORCE_RESULTS_STRICT

    if normalized_mode == "extended":
        return normalized_mode, False, BRUTE_FORCE_RESULTS_EXTENDED

    raise ValueError("mode must be 'strict' or 'extended'")


def get_target_expressions(
    results: dict,
    target: int,
) -> list[str]:
    """Return all expressions in `results` that evaluate to `target`."""
    return [
        expression
        for expression, value in results.items()
        if value == target
    ]


def print_all_expressions_query(
    target: int,
    mode: str,
    expressions: list[str],
    lim: int | None,
):
    """Print query output for all-expressions mode."""
    if not expressions:
        print(f"No expressions found for target={target} in {mode} mode.")
        return

    print()
    print((
        f"Total expressions that evaluate to {target} in "
        f"{mode} mode: {len(expressions)}"
    ))

    if lim is None or len(expressions) <= lim:
        for expression in expressions:
            print(expression)
        return

    print(f"Showing first {lim} expressions:")
    for expression in expressions[:lim]:
        print(expression)


def get_target_metadata(
    mode: str,
    strict_form: bool,
    target: int,
) -> dict:
    """Return target metadata from cache or compute it silently."""
    cache_key = (mode, target)
    cached_metadata = BRUTE_FORCE_TARGET_CACHE.get(cache_key)
    if cached_metadata is not None:
        return cached_metadata

    _, metadata = brute_force_alg(
        strict_form=strict_form,
        target=target,
        verbose=False,
        return_metadata=True,
    )
    return metadata


def print_target_query_stats(metadata: dict):
    """Print integer-result and evaluation counters for target queries."""
    print((
        "Evaluated expressions with integer result: "
        f"{metadata['expressions_with_integer_result']}"
    ))
    print((
        "Evaluated expressions with unique integer result: "
        f"{metadata['unique_integer_results']}"
    ))

    if metadata["target_found"]:
        print((
            "Expressions evaluated until first target hit (inclusive): "
            f"{metadata['evaluated_before_first_target']}"
        ))


def print_target_query_header(
    target: int,
    mode: str,
):
    """Print contextual header for target-specific query mode."""
    print_wrapper(f"""
        Target-specific brute-force query ({mode} form):
        Target value = {target}
    """)


def expressions_by_target(
    target,
    all_expressions: bool = True,
    lim: int | None = LIMIT,
    mode: str = "extended",
):
    """Query expressions by target value from brute-force results.

    Parameters
    ----------
    target : int
        Integer target to search for.
    all_expressions : bool
        True -> list all expressions (or first `lim`).
        False -> show first expression and evaluation stats until hit.
    lim : int | None
        Max number of expressions to print when `all_expressions=True`.
    mode : str
        "strict" or "extended".

    Returns
    -------
    list[str]
        Matching expressions according to selected query mode.
    """
    mode, strict_form, results = parse_query_mode(mode)

    if all_expressions:
        expressions = get_target_expressions(results, target)
        print_all_expressions_query(
            target=target,
            mode=mode,
            expressions=expressions,
            lim=lim,
        )
        return expressions

    metadata = get_target_metadata(
        mode=mode,
        strict_form=strict_form,
        target=target,
    )

    print_target_query_header(target=target, mode=mode)
    print_target_query_stats(metadata)
    print_target_status(metadata)

    if metadata["target_found"]:
        return [metadata["target_expression"]]

    return []


# Examples requested in the notebook statement.
EXPRESSIONS_3_STRICT = expressions_by_target(
    3,
    all_expressions=True,
    mode="strict",
)
EXPRESSIONS_4_EXTENDED = expressions_by_target(
    4,
    all_expressions=True,
    mode="extended",
)
EXPRESSION_42_EXTENDED = expressions_by_target(
    42,
    all_expressions=False,
    mode="extended",
)
EXPRESSION_42_STRICT = expressions_by_target(
    42,
    all_expressions=False,
    mode="strict",
)


Total expressions that evaluate to 3 in strict mode: 1952
Showing first 10 expressions:
3/1+8-2*4
3/1+8-4*2
4/1+5-2*3
4/1+5-3*2
4/1+9-2*5
4/1+9-5*2
4/2+6-1*5
4/2+6-5*1
4/2+7-1*6
4/2+7-6*1

Total expressions that evaluate to 4 in extended mode: 2208
Showing first 10 expressions:
3/1+9-2*4
3/1+9-4*2
4/1+6-2*3
4/1+6-3*2
4/2+5-1*3
4/2+5-3*1
4/2+7-1*5
4/2+7-5*1
4/2+8-1*6
4/2+8-6*1

Target-specific brute-force query (extended form):
Target value = 42
Evaluated expressions with integer result: 3238
Evaluated expressions with unique integer result: 134
Expressions evaluated until first target hit (inclusive): 18568
First expression found for target 42: 3/1+5*9-6

Target-specific brute-force query (strict form):
Target value = 42
Evaluated expressions with integer result: 3238
Evaluated expressions with unique integer result: 134
Expressions evaluated until first target hit (inclusive): 18568
First expression found for target 42: 3/1+5*9-6


En la salida de la celda anterior, para `target=42`, el conteo `Expressions evaluated until first target hit (inclusive)` puede aparecer igual en modo estricto y extendido. Esto ocurre cuando la primera expresión que alcanza `42` pertenece al conjunto estricto y se encuentra en la parte del recorrido que ambos modos comparten. El modo extendido solo añade ramas extra (expresiones que empiezan con `-` unario), pero esas ramas no llegan a explorarse si el objetivo ya fue encontrado antes.

### Diseño de un mejor algoritmo

8. (\*) Diseña un algoritmo que mejore la complejidad del algoritmo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta.

### Backtracking con poda fuerte exacta

La implementación queda separada en dos niveles:

- **Funciones auxiliares externas** (reutilizables):
  `normalize_figures`, `normalize_signs`, `suffix_transforms`,
  `can_reach_target` y `apply_operation`.
- **Función orquestadora**: `solve_with_backtracking`, que contiene la lógica
  de recorrido y decisión para modo estricto/extendido.

Dentro de `solve_with_backtracking`, el flujo se organiza con funciones anidadas
más pequeñas:

- `visit_leaf`: evalúa y registra hojas completas.
- `iter_next_steps`: genera transiciones de un nivel de DFS.
- `dfs_all`: recorre todo el árbol cuando `target=None`.
- `dfs_target`: aplica poda fuerte exacta cuando hay `target`.
- `run_from_root`: decide ruta (`dfs_all` o `dfs_target`) desde una raíz.
- `run_strict_root` y `run_extended_root`: inicializan las ramas raíz.
- `print_backtracking_summary` y `print_backtracking_target_status`: imprimen
  resultados, reutilizando utilidades de impresión de la celda anterior
  (`print_wrapper` y `print_target_status`).

La evaluación mantiene precedencia aritmética con estado incremental
`(total, term)` usando `Fraction`:

- `+` y `-` cierran el término actual (`total <- total + term`) y crean
  un nuevo `term`.
- `*` y `/` actualizan solo `term`.
- En hoja, el valor final es `total + term`.

Se puede ver una representación de como se desarrollaría cada una de las
posibilidades en este flujo en el siguiente diagrama:

                                       1
                          ┌────────────┼────────────┬────────────┐
                         [+]          [-]          [*]          [/]
                         / \           ...           ...           ...
                       2     6
                     /   \
                   [-]   [*] ...
                  /   \
                 3    ...
                  \
                   [*]
                 /  |  \
                4   5   6
               /    |    \
             [/]    ...    ...
            / | \
           5  6  7  ...

Para la ruta con `target`, la poda fuerte se aplica así:

1. Para cada estado parcial, se verifica si todavía es posible llegar al
   objetivo.
2. El sufijo restante se modela como transformación afín exacta:

   `valor_final = total + a * term + b`

3. Los pares `(a, b)` se precomputan con memoización en
   `suffix_transforms`, indexados por el estado restante (`rem_figures`,
   `rem_signs`) representado como tuplas.
4. Si no existe ningún par que satisfaga la igualdad con el `target`, la
   rama se poda inmediatamente.

Para la ruta `target=None`, no hay poda por objetivo, pero sí optimizaciones de
implementación (sin cambiar el orden asintótico):

- **Estado restante como tuplas inmutables**: para mantener compatibilidad con caché.
- **DFS especializada para enumeración**: se separa la ruta de enumeración (`dfs_all`) de la ruta con objetivo (`dfs_target`). Esto evita evaluar, en cada nodo, condiciones de poda que no aplican cuando `target=None`.
- **Mutación in-place de tokens**: la expresión parcial se guarda en un buffer preasignado y se escribe por posición. Así se evita crear listas nuevas con `append/pop` continuos, se reducen asignaciones y la presión de memoria.
- **Conversión reutilizable de cifras**: se precalculan `DIGIT_STR` y `DIGIT_FRACTION` para no reconstruir `str(d)` ni `Fraction(d)` en cada expansión del árbol.

Estas mejoras reducen el factor constante en la práctica para la enumeración
completa, manteniendo exactamente el mismo número de expresiones evaluadas.

Por qué se mantuvieron funciones anidadas dentro de `solve_with_backtracking`:

1. **Encapsulación**: esas funciones son detalles internos del algoritmo y no
   forman parte de la API pública.
2. **Closure / estado compartido**: permiten usar directamente `state`,
   `target_fraction`, `available_figures` y `available_signs` sin pasar una
   lista larga de parámetros en cada llamada.
3. **Flujo de control más simple**: cortar la búsqueda al encontrar objetivo
   (`state['target_found']`) queda más directo en la recursión.
4. **Menos argumentos y menos verbosidad**: reduce ruido de firma y mejora
   legibilidad del código principal.

#### Ejemplo sencillo de funcionamiento (con poda fuerte exacta)

Supongamos un estado parcial del DFS (ejemplo reducido para entender la idea):

Aquí, `total` es la suma acumulada de términos ya cerrados y `term` es el
término activo que todavía puede modificarse con `*` o `/` antes de
cerrarse con `+` o `-`.

- `total = 3`
- `term = 2`
- `rem_figures = (4, 5)`
- `rem_signs = ('*', '+')`

La función `suffix_transforms(rem_figures, rem_signs)` calcula todas las
transformaciones afines posibles del sufijo en la forma:

`valor_final = total + a * term + b`

En este ejemplo, las combinaciones válidas del sufijo producen:

1. `*4` y luego `+5` -> `(a, b) = (4, 5)`
2. `*5` y luego `+4` -> `(a, b) = (5, 4)`
3. `+4` y luego `*5` -> `(a, b) = (1, 20)`
   Porque `+4` cierra el `term` actual y crea un nuevo término `4`; luego
   `*5` solo afecta ese término nuevo: `valor_final = total + term + (4*5)`
   `= total + (1)*term + 20`.
4. `+5` y luego `*4` -> también `(a, b) = (1, 20)` (ya estaba)

Por tanto, `suffix_transforms(...)` devuelve (sin duplicados):

`{(4, 5), (5, 4), (1, 20)}`

Ahora `can_reach_target(total, term, rem_figures, rem_signs, target)` prueba
si existe algún par `(a, b)` que cumpla la igualdad exacta.

- Si `target = 16`:
  `3 + 4*2 + 5 = 16` -> **sí existe** solución de sufijo, no se poda.
- Si `target = 15`:
  `3 + 4*2 + 5 = 16`, `3 + 5*2 + 4 = 17`, `3 + 1*2 + 20 = 25`
  -> **no existe** solución de sufijo, la rama se poda inmediatamente.

Este ejemplo muestra exactamente qué hacen ambas funciones:

- `suffix_transforms` enumera y memoiza la factibilidad algebraica del sufijo.
- `can_reach_target` usa esa información para decidir poda exacta en cada nodo.


In [ ]:
import time
from fractions import Fraction
from functools import lru_cache


# Fixed operator order used by recursion.
OPS_ORDER = ('+', '-', '*', '/')
# Fixed digit order used in both strict and extended modes.
DIGITS_ORDER = tuple(range(1, 10))
# Precomputed string/fraction representations to avoid repeated conversions.
DIGIT_STR = {digit: str(digit) for digit in DIGITS_ORDER}
DIGIT_FRACTION = {digit: Fraction(digit) for digit in DIGITS_ORDER}


def normalize_figures(figures):
    """Return available digits as a sorted tuple without duplicates."""
    return tuple(sorted(set(figures)))


def normalize_signs(signs):
    """Return available operators in the fixed exploration order."""
    allowed = set(signs)
    return tuple(op for op in OPS_ORDER if op in allowed)


@lru_cache(maxsize=None)
def suffix_transforms(
    rem_figures,
    rem_signs,
):
    """Return affine transforms (a, b) with final = total + a*term + b.

    The cache key is the pair of remaining tuples (digits, operators). This
    allows exact feasibility checks for target search without recomputing
    suffix expansions repeatedly.
    """
    if not rem_signs:
        # No operators left: suffix contributes exactly the current term.
        return ((Fraction(1), Fraction(0)),)

    transforms = set()
    for sign_idx, op in enumerate(rem_signs):
        next_signs = rem_signs[:sign_idx] + rem_signs[sign_idx + 1:]

        for fig_idx, fig in enumerate(rem_figures):
            next_figures = rem_figures[:fig_idx] + rem_figures[
                fig_idx + 1:
            ]
            fig_fraction = DIGIT_FRACTION[fig]

            for a, b in suffix_transforms(next_figures, next_signs):
                # Compose transforms according to operator semantics.
                if op == '+':
                    transforms.add((Fraction(1), a * fig_fraction + b))
                elif op == '-':
                    transforms.add((Fraction(1), -a * fig_fraction + b))
                elif op == '*':
                    transforms.add((a * fig_fraction, b))
                else:  # op == '/'
                    transforms.add((a / fig_fraction, b))

    return tuple(transforms)


def can_reach_target(
    total,
    term,
    rem_figures,
    rem_signs,
    target_fraction,
):
    """Check whether any suffix transform can still hit target exactly."""
    for a, b in suffix_transforms(rem_figures, rem_signs):
        if total + a * term + b == target_fraction:
            return True
    return False


def apply_operation(
    total,
    term,
    sign,
    fig_fraction,
):
    """Apply operator with precedence-aware state update.

    The pair (total, term) encodes expression value as total + term.
    + / - close the current term; * / extend it.
    """
    if sign == '+':
        return total + term, fig_fraction
    if sign == '-':
        return total + term, -fig_fraction
    if sign == '*':
        return total, term * fig_fraction
    return total, term / fig_fraction


def solve_with_backtracking(  # NOSONAR(S3776)
    figures,
    signs,
    target=None,
    strict_form=False,
):
    """Solve expression search with backtracking (strict or extended mode).

    Modes
    -----
    - strict: only d1 s1 d2 s2 d3 s3 d4 s4 d5
    - extended: strict + unary-minus starting branch

    Behavior
    --------
    - target is None: full enumeration with optimized DFS (`dfs_all`).
    - target provided: target search with exact feasibility pruning
      (`dfs_target`).
    """
    mode = get_mode_name(strict_form)
    start = time.time()

    available_figures = normalize_figures(figures)
    available_signs = normalize_signs(signs)
    # Extended unary-minus branch can only use +, *, / afterwards.
    extended_signs = tuple(
        sign
        for sign in available_signs
        if sign != '-'
    )
    target_fraction = Fraction(target) if target is not None else None

    state = {
        'counter': 0,
        'results': {},
        'target_found': False,
        'target_expression': '',
    }

    def visit_leaf(
        tokens,
        used_length,
        total,
        term,
    ):
        """Evaluate one complete expression and update counters/results."""
        final_value = total + term
        state['counter'] += 1

        if final_value.denominator != 1:
            return

        integer_result = final_value.numerator
        expression_str = ''.join(tokens[:used_length])
        state['results'][expression_str] = integer_result

        if target_fraction is None:
            return

        if final_value == target_fraction:
            state['target_found'] = True
            state['target_expression'] = expression_str

    def iter_next_steps(
        rem_figures,
        rem_signs,
        total,
        term,
    ):
        """Yield next recursive states for one DFS level."""
        for sign_idx, sign in enumerate(rem_signs):
            next_signs = rem_signs[:sign_idx] + rem_signs[sign_idx + 1:]

            for fig_idx, fig in enumerate(rem_figures):
                next_figures = rem_figures[:fig_idx] + rem_figures[
                    fig_idx + 1:
                ]
                next_total, next_term = apply_operation(
                    total,
                    term,
                    sign,
                    DIGIT_FRACTION[fig],
                )
                yield (
                    sign,
                    fig,
                    next_signs,
                    next_figures,
                    next_total,
                    next_term,
                )

    def dfs_all(
        tokens,
        position,
        total,
        term,
        rem_figures,
        rem_signs,
    ):
        """Enumerate all valid expressions (no target pruning)."""
        if not rem_signs:
            visit_leaf(tokens, position, total, term)
            return

        for (
            sign,
            fig,
            next_signs,
            next_figures,
            next_total,
            next_term,
        ) in iter_next_steps(rem_figures, rem_signs, total, term):
            # In-place token mutation: overwrite operator slot.
            tokens[position] = sign
            # In-place token mutation: overwrite digit slot.
            tokens[position + 1] = DIGIT_STR[fig]
            dfs_all(
                tokens,
                position + 2,
                next_total,
                next_term,
                next_figures,
                next_signs,
            )

    def dfs_target(
        tokens,
        position,
        total,
        term,
        rem_figures,
        rem_signs,
    ):
        """Search target with exact feasibility pruning on remaining suffix."""
        if state['target_found']:
            return

        # Strong pruning: drop branch if no suffix can reach target.
        if not can_reach_target(
            total,
            term,
            rem_figures,
            rem_signs,
            target_fraction,
        ):
            return

        if not rem_signs:
            visit_leaf(tokens, position, total, term)
            return

        for (
            sign,
            fig,
            next_signs,
            next_figures,
            next_total,
            next_term,
        ) in iter_next_steps(rem_figures, rem_signs, total, term):
            # In-place token mutation: overwrite operator slot.
            tokens[position] = sign
            # In-place token mutation: overwrite digit slot.
            tokens[position + 1] = DIGIT_STR[fig]
            dfs_target(
                tokens,
                position + 2,
                next_total,
                next_term,
                next_figures,
                next_signs,
            )
            if state['target_found']:
                return

    def run_from_root(
        tokens,
        position,
        initial_term,
        rem_figures,
        rem_signs,
    ):
        """Dispatch DFS route (all-results vs target-specific)."""
        if target_fraction is None:
            dfs_all(
                tokens,
                position,
                Fraction(0),
                initial_term,
                rem_figures,
                rem_signs,
            )
            return

        dfs_target(
            tokens,
            position,
            Fraction(0),
            initial_term,
            rem_figures,
            rem_signs,
        )

    def run_strict_root():
        """Start strict-mode trees from each possible first digit."""
        # Preallocated shared token buffer (strict root).
        tokens = [''] * 9

        for fig_idx, first_fig in enumerate(available_figures):
            rem_figures = available_figures[:fig_idx] + available_figures[
                fig_idx + 1:
            ]
            # In-place token mutation: set strict root digit.
            tokens[0] = DIGIT_STR[first_fig]
            run_from_root(
                tokens,
                1,
                DIGIT_FRACTION[first_fig],
                rem_figures,
                available_signs,
            )
            if state['target_found']:
                return

    def run_extended_root():
        """Start extended unary-minus trees from each possible first digit."""
        if '-' not in available_signs:
            return

        # Preallocated shared token buffer (extended root).
        tokens = [''] * 8
        # In-place token mutation: set unary-minus prefix.
        tokens[0] = '-'

        for fig_idx, first_fig in enumerate(available_figures):
            rem_figures = available_figures[:fig_idx] + available_figures[
                fig_idx + 1:
            ]
            # In-place token mutation: set first digit after unary minus.
            tokens[1] = DIGIT_STR[first_fig]
            run_from_root(
                tokens,
                2,
                -DIGIT_FRACTION[first_fig],
                rem_figures,
                extended_signs,
            )
            if state['target_found']:
                return

    def print_backtracking_summary(run_time):
        """Print backtracking summary with the shared print wrapper."""
        integer_results = len(state['results'])
        unique_integer_values = len(set(state['results'].values()))
        print_wrapper(f"""
            Backtracking algorithm results ({mode} form):
            Run time = {run_time:.3f} seconds
            Total expressions evaluated: {state['counter']}
            Total expressions with integer result: {integer_results}
            Total unique integer results found: {unique_integer_values}
        """)

    def print_backtracking_target_status():
        """Reuse shared target-status printer with backtracking state."""
        if target is None:
            return

        target_metadata = {
            'target': target,
            'target_found': state['target_found'],
            'target_expression': state['target_expression'],
        }
        print_target_status(target_metadata)

    # Strict branch is always part of both modes.
    run_strict_root()

    # Extended mode adds unary-minus roots only when needed.
    if not strict_form and not state['target_found']:
        run_extended_root()

    run_time = time.time() - start
    print_backtracking_summary(run_time)
    print_backtracking_target_status()

    return state['results']


# Shared inputs for examples in this notebook section.
figures = list(FIGURES)
signs = list(SIGNS)

print('Finding a specific target (e.g., 42) - extended form')
solve_with_backtracking(
    figures,
    signs,
    42,
    strict_form=False,
)

print("\nFinding a specific target (e.g., 42) - strict form")
solve_with_backtracking(
    figures,
    signs,
    42,
    strict_form=True,
)

print("\nFinding all possible results - extended form")
all_results_extended = solve_with_backtracking(
    figures,
    signs,
    strict_form=False,
)

print("\nFinding all possible results - strict form")
all_results_strict = solve_with_backtracking(
    figures,
    signs,
    strict_form=True,
)

Finding a specific target (e.g., 42) - extended form

Backtracking algorithm results (extended form):
Run time = 0.175 seconds
Total expressions evaluated: 1
Total expressions with integer result: 1
Total unique integer results found: 1
First expression found for target 42: 1+5*9-8/2

Finding a specific target (e.g., 42) - strict form

Backtracking algorithm results (strict form):
Run time = 0.012 seconds
Total expressions evaluated: 1
Total expressions with integer result: 1
Total unique integer results found: 1
First expression found for target 42: 1+5*9-8/2

Finding all possible results - extended form

Backtracking algorithm results (extended form):
Run time = 0.980 seconds
Total expressions evaluated: 381024
Total expressions with integer result: 94824
Total unique integer results found: 148

Finding all possible results - strict form

Backtracking algorithm results (strict form):
Run time = 0.922 seconds
Total expressions evaluated: 362880
Total expressions with integer result: 9

En la ejecución mostrada, el backtracking con poda fuerte mejora claramente la búsqueda de objetivo (`target=42`) frente a fuerza bruta: encuentra una expresión válida tras apenas **1** evaluación, mientras que fuerza bruta en modo target necesita evaluar **3,238** expresiones para lograr el primer acierto.

Cuando se piden **todos los resultados** (`target=None`), ambos algoritmos deben recorrer el mismo número de hojas por modo (estricto/extendido). La diferencia práctica pasa a estar en la implementación: la versión actual de backtracking usa una ruta optimizada con DFS especializada, estado incremental `(total, term)` y mutación in-place del buffer de tokens, lo que reduce sobrecarga aunque el conteo de expresiones evaluadas sea el mismo.

### Complejidad

9. (\*) Calcula la complejidad del algoritmo.

La complejidad depende del modo de uso.

Sea:

- $d$: número de cifras disponibles,
- $s$: número de signos disponibles,
- $m$: número de operadores por expresión (aquí, $m=4$).

Para **enumeración completa** (`target=None`), no hay poda por objetivo y se recorre todo el espacio:

- **Modo estricto**:
  $$
  T_{strict}=\Theta(P(s,m)\cdot P(d,m+1))
  $$
- **Modo extendido**:
  $$
  T_{extended}=\Theta(P(s,m)\cdot P(d,m+1)+P(s-1,m-1)\cdot P(d,m))
  $$

En esta instancia:

- estricto: $\Theta(362{,}880)$,
- extendido: $\Theta(381{,}024)$.

La optimización de implementación (separación `dfs_all`/`dfs_target`, buffer in-place y conversiones precalculadas) **no cambia** ese orden asintótico, pero sí reduce el factor constante.

Para **búsqueda de objetivo** (`target` definido), se activa poda fuerte exacta (`can_reach_target` + `suffix_transforms`):

- en promedio, reduce mucho el árbol explorado,
- en peor caso, sigue acotado por el recorrido completo del modo.

Por tanto:

$$
T_{mode,target}=O(T_{mode})
$$

con mejora práctica significativa cuando la poda es efectiva.

Memoria:

- $O(k)$ para el diccionario de expresiones enteras almacenadas,
- más la caché de transformaciones del sufijo (por estados de tuplas de cifras/signos restantes).

### Datos de entrada aleatorios

10. Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios.


In [5]:
import numpy as np


# Target sampling interval derived from strict-mode integer range.
MIN_VALUE = -69
MAX_VALUE = 77
# Random generator
g = np.random.default_rng(42)


def generate_random_input(size=LIMIT, low=MIN_VALUE, high=MAX_VALUE):
    """Generate random integer targets in [low, high]."""
    # integers upper bound is exclusive, so use high+1.
    return g.integers(low, high+1, size)


print(generate_random_input())

[-56  44  27  -5  -6  57 -57  33 -40 -56]


### Aplicación del algoritmo

11. Aplica el algoritmo al juego de datos generado.


In [6]:
def compare_random_targets_with_bruteforce_and_backtracking(
        strict_form=False,
        size=LIMIT):
    """Compare brute force vs backtracking for random targets.

    For each sampled target, this function prints both algorithm summaries
    with the same mode (strict or extended) to make the comparison direct.
    """
    mode = "strict" if strict_form else "extended"
    targets = [int(value) for value in generate_random_input(size=size)]

    print(f"Random targets ({mode} form): {targets}")

    for i, target in enumerate(targets, start=1):
        print()
        print("=" * 72)
        print(f"Target {i}/{len(targets)} = {target} ({mode} form)")

        print()
        print("[Brute force]")
        brute_force_alg(strict_form=strict_form, target=target)

        print()
        print("[Backtracking]")
        solve_with_backtracking(
            figures, signs, target,
            strict_form=strict_form)


# Run random-target comparison in extended mode by default.
compare_random_targets_with_bruteforce_and_backtracking(strict_form=False)

Random targets (extended form): [8, 74, 39, 42, 36, 46, 6, -51, 54, -3]

Target 1/10 = 8 (extended form)

[Brute force]

Brute-force algorithm results (extended form):
Run time = 0.024 seconds
Evaluated expressions until first target hit (inclusive) = 5431
Evaluated expressions with integer result = 811
Evaluated expressions with unique integer result = 76
Minimum value = 4/2+1-8*9 = -69
Maximum value = 4/2+9-1*3 = 8
First expression found for target 8: 4/2+9-1*3

[Backtracking]

Backtracking algorithm results (extended form):
Run time = 0.002 seconds
Total expressions evaluated: 1
Total expressions with integer result: 1
Total unique integer results found: 1
First expression found for target 8: 1+2*5-9/3

Target 2/10 = 74 (extended form)

[Brute force]

Brute-force algorithm results (extended form):
Run time = 0.091 seconds
Evaluated expressions until first target hit (inclusive) = 20336
Evaluated expressions with integer result = 3536
Evaluated expressions with unique integer result 

### Comparación en targets aleatorios

En la celda anterior se comparan, para cada `target` aleatorio generado, los resultados de **fuerza bruta** y **backtracking con poda fuerte** en el mismo modo (por defecto, modo extendido). La comparación muestra dos patrones consistentes:

- Ambos algoritmos encuentran expresiones válidas para los objetivos del muestreo cuando el objetivo pertenece al rango alcanzable.
- Si el target pertenece al rango alcanzable, para **todos los casos**, el backtracking **siempre** encontrará una expresión objetivo al **primer intento** (`Total expressions evaluated = 1`), debido a que la poda de factibilidad elimina todas las ramas que no pueden alcanzar el objetivo, y al final la única expresión completa que se evalúa es la que produce el `target`, dado que todas las demás ramas se podan antes de llegar a hoja (`Total expressions evaluated` solo cuanta expresiones **completas** evaluadas).
- El algoritmo por fuerza bruta puede necesitar evaluar miles de expresiones antes de encontrar la primera que alcance el `target`, dependiendo del orden de generación y la posición del `target` en el espacio de soluciones.

Este comportamiento ocurre principalmente por tres motivos:

1. La poda de factibilidad (`can_reach_target`) es exacta: solo se exploran ramas cuyo sufijo todavía puede llegar al objetivo.
2. El contador de expresiones se incrementa solo al evaluar hojas completas; muchas ramas se descartan antes de llegar a hoja y no cuentan como expresiones evaluadas.
3. El orden de exploración, combinado con la alta cobertura de objetivos alcanzables en el rango del problema, hace que con frecuencia la primera rama factible ya produzca el `target`.

La mejora no es una garantía universal para todos los objetivos ni para cualquier orden de recorrido, pero en las ejecuciones del notebook se observa una ventaja práctica clara frente a fuerza bruta en búsqueda de objetivo. Además, el tiempo de ejecución de ambos algoritmos está parcialmente optimizado por el uso de caché: en fuerza bruta mediante `BRUTE_FORCE_TARGET_CACHE` para reutilizar búsquedas por target, y en backtracking mediante `lru_cache` en `suffix_transforms` para reutilizar transformaciones de sufijos.

### Referencias

12. Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo.
    - Universidad Internacional de Valencia (VIU). (2025). Apuntes y recursos de la asignatura Algoritmos de Optimización. Material docente en el aula virtual.
    - Universidad Internacional de Valencia (VIU). (2025). Apuntes y recursos de la asignatura Python para la Inteligencia Artificial. Material docente en el aula virtual.
    - https://docs.python.org/3/library/itertools.html

### Estudio avanzado del problema

13. Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño.

Para avanzar en el estudio del problema, las siguientes líneas son especialmente relevantes:

- Escalado del problema y generalización.
  Analizar qué ocurre al aumentar el número de cifras disponibles, el número de operadores o permitir repetición controlada. Esto permite estudiar cómo cambian el espacio de búsqueda y el rendimiento de cada estrategia.

- Ordenación heurística de ramas.
  Priorizar operadores/cifras que históricamente acercan más rápido al objetivo puede reducir aún más el tiempo medio en búsquedas con `target`.

- Poda adicional para enumeración completa.
  Aunque en `target=None` no se puede evitar visitar todo el conjunto de soluciones, sí se puede seguir reduciendo coste constante con mejoras de implementación (estructuras más compactas, menor creación de objetos, recorrido más cache-friendly y paralelización por ramas raíz).

- Variantes del objetivo.
  Reformular el problema para optimizar criterios como: minimizar número de operaciones, minimizar magnitud de cifras usadas, o maximizar cobertura de un intervalo bajo restricciones adicionales.

- Evaluación experimental sistemática.
  Diseñar benchmarks reproducibles (mismo conjunto de targets, semillas fijas, varias corridas) para comparar robustamente fuerza bruta y backtracking en tiempo, expresiones evaluadas y estabilidad.
